File: Lab04/trapecioPool.go
package main

import (
	"fmt"      // Para impresión en consola
	"runtime"  // Para detectar el número de núcleos disponibles
	"sync"     // Para sincronización con WaitGroup y Mutex
	"time"     // Para medir tiempos de ejecución
)

// ===============================================
// DEFINICIÓN DE UN THREAD POOL PERSONALIZADO
// ===============================================
type ThreadPool struct {
	jobs    chan func()     // Canal para recibir funciones a ejecutar
	wg      sync.WaitGroup  // WaitGroup para esperar que terminen los workers
	workers int             // Número de trabajadores (hilos)
}

// Constructor del ThreadPool
func NewThreadPool(workers int) *ThreadPool {
	pool := &ThreadPool{
		jobs:    make(chan func(), workers*10), // Canal con buffer
		workers: workers,
	}

	// Crear los workers como goroutines que esperan funciones del canal
	for i := 0; i < workers; i++ {
		pool.wg.Add(1)
		go func() {
			defer pool.wg.Done() // Marca que este worker ha terminado cuando sale
			for job := range pool.jobs {
				job() // Ejecuta la función recibida
			}
		}()
	}

	return pool
}

// Método para enviar un trabajo (función) al pool
func (p *ThreadPool) Submit(job func()) {
	p.jobs <- job
}

// Espera a que todos los workers terminen y cierra el canal
func (p *ThreadPool) Wait() {
	close(p.jobs)   // Cierra el canal de trabajos (los workers saldrán del bucle)
	p.wg.Wait()     // Espera a que todos los workers terminen
}

// ===============================================
// ESTRUCTURA PARA CALCULAR INTEGRALES POR TRAPECIO
// ===============================================
type TrapecioPool struct {
	funcion func(float64) float64 // Función a integrar
	a       float64               // Límite inferior
	b       float64               // Límite superior
}

// Constructor para TrapecioPool
func NewTrapecioPool(f func(float64) float64, a, b float64) *TrapecioPool {
	return &TrapecioPool{
		funcion: f,
		a:       a,
		b:       b,
	}
}

// ===============================================
// MÉTODO PARA CALCULAR EL ÁREA CON THREAD POOL
// ===============================================
func (tp *TrapecioPool) CalcularConPool(n int, pool *ThreadPool) float64 {
	h := (tp.b - tp.a) / float64(n)               // Tamaño de cada subintervalo
	numTareas := runtime.NumCPU() * 4             // Más tareas que núcleos para mejorar carga
	var wg sync.WaitGroup                         // WaitGroup para sincronizar las tareas
	resultados := make(chan float64, numTareas)   // Canal para recibir sumas parciales
	suma := 0.5 * (tp.funcion(tp.a) + tp.funcion(tp.b)) // Suma inicial con los extremos

	puntosPorTarea := n / numTareas // Cuántos puntos calcula cada tarea

	// Crear tareas y enviarlas al thread pool
	for tareaID := 0; tareaID < numTareas; tareaID++ {
		wg.Add(1)
		start := tareaID * puntosPorTarea + 1
		end := (tareaID + 1) * puntosPorTarea
		if tareaID == numTareas-1 {
			end = n - 1 // La última tarea puede tomar más puntos
		}

		startLocal := start
		endLocal := end

		// Enviamos una tarea al pool
		pool.Submit(func() {
			defer wg.Done()  // Marca tarea como completada
			sumaLocal := 0.0
			for i := startLocal; i <= endLocal; i++ {
				x := tp.a + float64(i)*h
				sumaLocal += tp.funcion(x)
			}
			resultados <- sumaLocal // Enviamos el resultado parcial al canal
		})
	}

	// Goroutine para cerrar el canal una vez terminadas todas las tareas
	go func() {
		wg.Wait()
		close(resultados)
	}()

	// Acumulamos todos los resultados recibidos desde las tareas
	for resultado := range resultados {
		suma += resultado
	}

	return suma * h // Multiplicamos por el tamaño de subintervalo
}

// ===============================================
// FUNCIÓN PRINCIPAL
// ===============================================
func main() {
	// Definimos la función a integrar
	funcion := func(x float64) float64 {
		return 2*x*x + 3*x + 0.5
	}

	// Detectamos cantidad de núcleos disponibles
	numCores := runtime.NumCPU()

	// Creamos el thread pool
	pool := NewThreadPool(numCores)
	defer pool.Wait() // Esperamos que todos los trabajos terminen al final

	// Creamos la calculadora de integrales
	calculadora := NewTrapecioPool(funcion, 2, 20)

	// Información básica
	fmt.Printf("Usando Thread Pool con %d threads\n", numCores)
	fmt.Println("==============================================")

	// Ejecutamos para diferentes cantidades de divisiones (n)
	for n := 10; n <= 1000000; n *= 10 {
		start := time.Now() // Tiempo de inicio
		area := calculadora.CalcularConPool(n, pool)
		tiempo := time.Since(start) // Tiempo transcurrido

		// Mostramos resultado y tiempo en milisegundos
		fmt.Printf("n=%d: Área=%.6f, Tiempo=%.3f ms\n", 
			n, area, float64(tiempo.Nanoseconds())/1e6)
	}
}

File: Lab04/TrapecioPool.java
import java.util.concurrent.*; // Importa las clases necesarias para la concurrencia y las tareas en segundo plano.
import java.util.function.Function; // Importa la clase Function que se utiliza para definir la función matemática a integrar.

public class TrapecioPool {
    // Definimos el tamaño del pool de hilos (threads) según la cantidad de procesadores disponibles en la máquina.
    private static final int POOL_SIZE = Runtime.getRuntime().availableProcessors();
    // Creamos un thread pool con el tamaño definido anteriormente.
    private static final ExecutorService threadPool = Executors.newFixedThreadPool(POOL_SIZE);
    
    private Function<Double, Double> funcion; // La función que queremos integrar.
    private double a; // Límite inferior de la integral.
    private double b; // Límite superior de la integral.

    // Constructor de la clase que recibe la función y los límites de integración.
    public TrapecioPool(Function<Double, Double> funcion, double a, double b) {
        this.funcion = funcion;
        this.a = a;
        this.b = b;
    }

    // Método que realiza la integración usando el método de los trapecios y un thread pool para paralelizar el trabajo.
    public double calcularConPool(int n) throws InterruptedException, ExecutionException {
        // Calculamos el tamaño de cada subintervalo
        double h = (b - a) / n;
        // La suma inicial es la mitad de los valores de los extremos (puntos a y b)
        double suma = 0.5 * (funcion.apply(a) + funcion.apply(b));
        
        // Definimos cuántas tareas vamos a dividir. En este caso, usamos el pool de hilos con más tareas que hilos,
        // lo que permite mejor distribución de trabajo entre los hilos.
        int tareas = POOL_SIZE * 4; 
        // Cada tarea calculará una cantidad proporcional de puntos.
        int puntosPorTarea = n / tareas;
        
        // Array para almacenar las tareas que se envían al thread pool.
        Future<Double>[] resultados = new Future[tareas];
        
        // Dividimos el trabajo en subintervalos y enviamos tareas al thread pool.
        for (int t = 0; t < tareas; t++) {
            // Cada tarea calcula una parte del intervalo
            final int start = t * puntosPorTarea + 1;
            final int end = (t == tareas - 1) ? n - 1 : (t + 1) * puntosPorTarea;
            
            // Enviamos la tarea al thread pool para su ejecución en paralelo
            resultados[t] = threadPool.submit(() -> {
                double sumaParcial = 0;
                // Cada hilo calcula una suma parcial de la integral sobre un subintervalo.
                for (int i = start; i <= end; i++) {
                    double x = a + i * h;
                    sumaParcial += funcion.apply(x);
                }
                return sumaParcial;
            });
        }
        
        // Esperamos a que todas las tareas terminen y sumamos sus resultados parciales.
        for (Future<Double> resultado : resultados) {
            suma += resultado.get(); // Espera a que la tarea termine y obtiene el resultado.
        }
        
        // Multiplicamos la suma total por el tamaño del intervalo (h) para obtener el valor de la integral.
        return suma * h;
    }
    
    // Método para cerrar el pool de hilos una vez terminado el cálculo.
    public static void shutdownPool() {
        threadPool.shutdown(); // Cierra el thread pool.
    }

    // Método principal para probar la clase y calcular la integral para diferentes valores de n.
    public static void main(String[] args) {
        // Definimos la función a integrar: 2x^2 + 3x + 0.5
        Function<Double, Double> funcion = x -> 2 * Math.pow(x, 2) + 3 * x + 0.5;
        // Creamos una instancia de la clase TrapecioPool con la función y los límites de integración.
        TrapecioPool calculadora = new TrapecioPool(funcion, 2, 20);
        
        try {
            // Imprimimos información sobre la cantidad de hilos en el pool.
            System.out.println("Usando Thread Pool con " + POOL_SIZE + " threads");
            System.out.println("==============================================");
            
            // Iteramos sobre diferentes valores de n para ver cómo cambia el tiempo de ejecución y el resultado de la integral.
            for (int n = 10; n <= 1000000; n *= 10) {
                long startTime = System.nanoTime(); // Marca el tiempo de inicio.
                double area = calculadora.calcularConPool(n); // Calculamos el área usando el método de Trapecio con thread pool.
                long endTime = System.nanoTime(); // Marca el tiempo de finalización.
                
                // Imprimimos el resultado con el tiempo de ejecución.
                System.out.printf("n=%d: Área=%.6f, Tiempo=%.3f ms\n", 
                    n, area, (endTime - startTime) / 1e6);
            }
            
        } catch (Exception e) {
            // Si ocurre un error, lo mostramos.
            e.printStackTrace();
        } finally {
            // Cerramos el thread pool al finalizar el cálculo.
            shutdownPool();
        }
    }
}
